---
title: "Investigating Pertussis Resurgence"
author: "Jason Hsiao (PID: A15871650)"
format: pdf
---

# Investigating Pertussis Resurgence

Libraries Used:
```{r}
suppressPackageStartupMessages({
library(datapasta)
library(ggplot2)
library(jsonlite)
library(lubridate)
library(dplyr)
library(readxl)
})

```

## Pertussis by Year - CDC Data 
https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html#print 

```{r, echo = FALSE}
# Using Datapasta to Paste Dataframe 
cdc <- data.frame(
                    Year = c(1922L,
                                   1923L,1924L,1925L,1926L,1927L,1928L,
                                   1929L,1930L,1931L,1932L,1933L,1934L,1935L,
                                   1936L,1937L,1938L,1939L,1940L,1941L,
                                   1942L,1943L,1944L,1945L,1946L,1947L,1948L,
                                   1949L,1950L,1951L,1952L,1953L,1954L,
                                   1955L,1956L,1957L,1958L,1959L,1960L,
                                   1961L,1962L,1963L,1964L,1965L,1966L,1967L,
                                   1968L,1969L,1970L,1971L,1972L,1973L,
                                   1974L,1975L,1976L,1977L,1978L,1979L,1980L,
                                   1981L,1982L,1983L,1984L,1985L,1986L,
                                   1987L,1988L,1989L,1990L,1991L,1992L,1993L,
                                   1994L,1995L,1996L,1997L,1998L,1999L,
                                   2000L,2001L,2002L,2003L,2004L,2005L,
                                   2006L,2007L,2008L,2009L,2010L,2011L,2012L,
                                   2013L,2014L,2015L,2016L,2017L,2018L,
                                   2019L,2020L,2021L),
                    Cases = c(107473,
                                   164191,165418,152003,202210,181411,
                                   161799,197371,166914,172559,215343,179135,
                                   265269,180518,147237,214652,227319,103188,
                                   183866,222202,191383,191890,109873,
                                   133792,109860,156517,74715,69479,120718,
                                   68687,45030,37129,60886,62786,31732,28295,
                                   32148,40005,14809,11468,17749,17135,
                                   13005,6799,7717,9718,4810,3285,4249,
                                   3036,3287,1759,2402,1738,1010,2177,2063,
                                   1623,1730,1248,1895,2463,2276,3589,
                                   4195,2823,3450,4157,4570,2719,4083,6586,
                                   4617,5137,7796,6564,7405,7298,7867,
                                   7580,9771,11647,25827,25616,15632,10454,
                                   13278,16858,27550,18719,48277,28639,
                                   32971,20762,17972,18975,15609,18617,6124,
                                   2116)
)

```

Let's visualize it!
```{r}
# Visualizing
ggplot(cdc, aes(Year, Cases)) + 
  geom_point() +
  geom_line() +
  theme_bw() + 
  # Adding vlines for wP and aP vaccination rollouts
  geom_vline(xintercept = c(1946, 1996, 2019), color = c("blue", "red","darkgreen"), linetype = "dotted") +
  ggtitle("Pertussis Cases by Year")

```

## Exploring CMI-PB Data
```{r}
# Reading json file and assigning to object 'subject'
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
# Determining Number in Dataset
summary(subject)
# Determining aP vs wP numbers in subject object
table(subject$infancy_vac)
# By sex
table(subject$biological_sex)
# By race and biological sex
table(subject$race, subject$biological_sex)
# Average age of wP and aP individuals
subject$age <- time_length(today() - ymd(subject$year_of_birth), "year")
subject$age # overall ages

ap <- subject %>% filter(infancy_vac == "aP")
round(summary(ap$age)) # ages of aP vaccinated

wp <- subject %>% filter(infancy_vac == "wP")
round(summary(wp$age)) # ages of wP vaccinated

# Age of all individuals at time of boost
subject$age_at_boost <- time_length(ymd(subject$date_of_boost) - ymd(subject$year_of_birth), "year")
subject$age_at_boost
# Plotting aP vs wP ages at time of boost
ggplot(subject) +
  aes(time_length(age_at_boost),
  fill = as.factor(infancy_vac)) +
  geom_histogram(show.legend = FALSE) +
  facet_wrap(vars(infancy_vac), nrow = 2) +
  xlab("Age at Boost (Years)") +
  ylab("Frequency")

```

## Merging subject and specimen data 
```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) # Importing specimen data
titer <- read_json("https://www.cmi-pb.org/api/plasma_ab_titer", simplifyVector = TRUE) # Importing titer data

# Matching subject information (metadata) to specimen id to form a new dataframe 
meta <- inner_join(specimen, subject, by = "subject_id")
dim(meta)
head(meta)

# Adding antibody titer (collected in specimens) information to meta
abdata <- inner_join(titer, meta, by = "specimen_id")
dim(abdata)
head(abdata)

# Number of specimens for each isotype of Ig in abdata 
table(abdata$isotype)
# Specimens by visit type
table(abdata$visit)


```

## Examining IgG1 Ab Titer Levels

We notice from above that visits 8-12 have considerable less samples, so we will remove those. Additionally, we want to look at IgG1 only:
```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", antigen =="PT", dataset == "2021_dataset")
head(ig1)

# Plotting ig1 (antigen-specific IgG1 levels by visit)
ggplot(ig1) +
  aes(planned_day_relative_to_boost, MFI_normalised, col = infancy_vac) +
  geom_line(aes(group = subject_id), alpha = 0.2) +
  geom_smooth(se = FALSE, span = 0.4, linewidth = 2) +
  geom_point(alpha = 0.2)

# For 2022 dataset?
ig2 <- abdata %>% filter(isotype == "IgG1", antigen =="PT", dataset == "2022_dataset")
head(ig2,15)

ggplot(ig2) +
  aes(planned_day_relative_to_boost, MFI_normalised, col = infancy_vac) +
  geom_line(aes(group = subject_id), alpha = 0.2) +
  geom_smooth(se = FALSE, span = 0.4, linewidth = 2) +
  geom_point(alpha = 0.2)

```

## CMI-PB RNAseq Data
```{r}
#Reading JSON file of RNAseq data, specifically for IghG1 gene
rna <- read_json("https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7", simplifyVector = TRUE) 

#Joining RNAseq data of specimens to specimen ID in meta
ssrna <- inner_join(rna, meta, by = "specimen_id")

#Visualizing reads of IghG1 by visit
ggplot(ssrna) +
  aes(visit, tpm, group = subject_id) +
  geom_point() +
  geom_line(alpha = 0.2)

#Are there differences by vaccination status? (Removing Visit 7: no aP data)
ssrna %>%
  filter(visit != 7) %>%
    ggplot() +
    aes(tpm, col = infancy_vac) +
    geom_boxplot() +
    facet_wrap(vars(visit), nrow = 2) 

```


